salt/modules/kubernetes.py

```diff
@@ -106,91 +98,36 @@ def _setup_conn(**kwargs):
-        kubernetes.client.configuration.api_key_prefix = {'authorization': __salt__['config.option']('kubernetes.api_key_prefix')}
-
-    if ca_cert_file:
-        kubernetes.client.configuration.ssl_ca_cert = ca_cert_file
-    elif ca_cert:
-        with tempfile.NamedTemporaryFile(prefix='salt-kube-', delete=False) as ca:
-            ca.write(base64.b64decode(ca_cert))
-            kubernetes.client.configuration.ssl_ca_cert = ca.name
-    else:
-        kubernetes.client.configuration.ssl_ca_cert = None
-
-    if client_cert_file:
-        kubernetes.client.configuration.cert_file = client_cert_file
-    elif client_cert:
-        with tempfile.NamedTemporaryFile(prefix='salt-kube-', delete=False) as c:
-            c.write(base64.b64decode(client_cert))
-            kubernetes.client.configuration.cert_file = c.name
-    else:
-        kubernetes.client.configuration.cert_file = None
-
-    if client_key_file:
-        kubernetes.client.configuration.key_file = client_key_file
-    elif client_key:
-        with tempfile.NamedTemporaryFile(prefix='salt-kube-', delete=False) as k:
-            k.write(base64.b64decode(client_key))
-            kubernetes.client.configuration.key_file = k.name
-    else:
-        kubernetes.client.configuration.key_file = None
+    if 'kubeconfig' in kwargs:
+        kubeconfig = kwargs.get('kubeconfig')
+    elif 'kubeconfig_data' in kwargs:
+        kubeconfig_data = kwargs.get('kubeconfig_data')
+        kubeconfig_data_overwrite = True
+
+    if 'context' in kwargs:
+        context = kwargs.get('context')
+
+    if (kubeconfig_data and not kubeconfig) or (kubeconfig_data and kubeconfig_data_overwrite):
+        with tempfile.NamedTemporaryFile(prefix='salt-kubeconfig-', delete=False) as kcfg:
+            kcfg.write(base64.b64decode(kubeconfig_data))
+            kubeconfig = kcfg.name
+    kubernetes.config.load_kube_config(config_file=kubeconfig, context=context)
 
     # The return makes unit testing easier
-    return vars(kubernetes.client.configuration)
+    return {'kubeconfig': kubeconfig, 'context': context}
 
 
 def _cleanup(**kwargs):
-    ca = kubernetes.client.configuration.ssl_ca_cert
-    cert = kubernetes.client.configuration.cert_file
-    key = kubernetes.client.configuration.key_file
-    if cert and os.path.exists(cert) and os.path.basename(cert).startswith('salt-kube-'):
-        salt.utils.files.safe_rm(cert)
-    if key and os.path.exists(key) and os.path.basename(key).startswith('salt-kube-'):
-        salt.utils.files.safe_rm(key)
-    if ca and os.path.exists(ca) and os.path.basename(ca).startswith('salt-kube-'):
-        salt.utils.files.safe_rm(ca)
+    if 'kubeconfig' in kwargs:
+        kubeconfig = kwargs.get('kubeconfig')
+        if os.path.exists(kubeconfig) and os.path.basename(kubeconfig).startswith('salt-kubeconfig-'):
+            salt.utils.files.safe_rm(kubeconfig)
```

Comment:
No need to `stat` file here, you could just `try/except` and check for `errno.ENOENT`:
```python
try:
    os.unlink(kubeconfig)
except (IOError, OSError) as err:
    if err.errno != errno.ENOENT:
        log.exception()
```

Reply:
BTW, interesting that you don't know what is inside `kubeconfig`... :wink:

Reply:
No, this is not working. I only want to remove it when it is a temp file created by the module itself.
So just unlink is not an option. I can either just check if I have something in kubeconfig or call
path.exists().  I like to check if kubeconfig is a "path" and if it exists before I remove it.